modules/shared/src/main/scala/org/tessellation/schema/swap.scala

```diff
@@ -0,0 +1,78 @@
+package org.tessellation.schema
+
+import org.tessellation.schema.address.Address
+import org.tessellation.security.hash.Hash
+
+import derevo.cats.{order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import eu.timepit.refined.types.numeric.PosLong
+import io.estatico.newtype.macros.newtype
+
+object swap {
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapAmount(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapSalt(value: Long)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class CurrencyId(value: Address)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendReference(value: Hash) // exact type TBD (e.g. lastRef?)
+
+  @derive(decoder, encoder, order, show)
+  case class AllowSpend(
+    source: Address,
+    destination: Address,
+    currency: CurrencyId,
+    amount: SwapAmount,
+    fee: AllowSpendFee,
+    parent: AllowSpendReference,
+    salt: SwapSalt,
+    lastValidOrdinal: SnapshotOrdinal,
+    approvers: List[Address]
+  )
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapAMMDataUpdateFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class PriceRange(min: SwapAmount, max: SwapAmount)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class PoolId(value: Address) // exact type TBD
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapReference(value: Address) // exact type TBD
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SpendTransactionFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class SpendTransaction(
+    fee: SpendTransactionFee,
+    lastValidOrdinal: SnapshotOrdinal,
+    allowSpendRef: Hash,
+    currency: Option[CurrencyId],
```

Comment:
so if there is no currency defined, it is a DAG spend transaction, correct?

Reply:
yes